lib/Dialect/FIRRTL/Transforms/Lint.cpp

```diff
@@ -0,0 +1,58 @@
+//===- Lint.cpp -------------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/ADT/APSInt.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct LintPass : public LintBase<LintPass> {
+  void runOnOperation() override {
+    auto fModule = getOperation();
+    (void)failableParallelForEach(
+        fModule.getContext(), fModule.getOps(),
+        [&](Operation &op) { return checkAssert(&op); });
+  };
+  LogicalResult checkAssert(Operation *op) {
+    Value predicate;
+    if (auto a = dyn_cast<AssertOp>(op)) {
+      if (auto constant = a.getEnable().getDefiningOp<firrtl::ConstantOp>())
+        if (constant.getValue().isOne()) {
+          predicate = a.getPredicate();
+        }
+    } else if (auto a = dyn_cast<VerifAssertIntrinsicOp>(op))
+      predicate = a.getProperty();
+
+    if (!predicate)
+      return success();
+    if (auto constant = predicate.getDefiningOp<firrtl::ConstantOp>())
+      if (constant.getValue().isZero())
+        return op->emitOpError(
+            "is guaranteed to fail simulation, as the predicate is "
+            "constant false");
+
+    if (auto reset = predicate.getDefiningOp<firrtl::AsUIntPrimOp>())
+      if (firrtl::type_isa<ResetType, AsyncResetType>(
+              reset.getInput().getType()))
+        return op->emitOpError("is guarunteed to fail simulation, as the "
+                               "predicate is a reset signal")
+                   .attachNote(reset.getLoc())
```

Comment:
I'm not sure what the location on the asUint cast will point to, because the user doesn't actually create these ops. Maybe they will have good locations, but maybe we could use the location from the operand Value to asUint, which should point to user code?

Reply:
I agree, better to track the reset signal location.